lib/improver/tests/blending/spatial_weights/test_SpatialWeightsForMissingData.py

```diff
@@ -0,0 +1,787 @@
+        self.assertArrayAlmostEqual(result.data, expected_result)
+        self.assertEqual(result.metadata, self.spatial_weights_cube.metadata)
+
+    @ManageWarnings(ignored_messages=[
+        "Collapsing a non-contiguous coordinate."])
+    def test_already_normalised(self):
+        """Test nothing happens if the input data is already normalised."""
+        self.spatial_weights_cube.data = np.array([[[[0.4, 0, 0.2],
+                                                     [0.4, 0, 0.2]],
+                                                    [[0.4, 0, 0.2],
+                                                     [0.4, 0, 0.2]]],
+                                                   [[[0, 0, 0.5],
+                                                     [0, 0, 0.5]],
+                                                    [[0, 0, 0.5],
+                                                     [0, 0, 0.5]]],
+                                                   [[[0.6, 1.0, 0.3],
+                                                     [0.6, 1.0, 0.3]],
+                                                    [[0.6, 1.0, 0.3],
+                                                     [0.6, 1.0, 0.3]]]],
+                                                  dtype=np.float32)
+        result = self.plugin.normalised_masked_weights(
+            self.spatial_weights_cube, "forecast_reference_time")
+        self.assertArrayAlmostEqual(
+            result.data, self.spatial_weights_cube.data)
+        self.assertEqual(result.metadata, self.spatial_weights_cube.metadata)
+
+    @ManageWarnings(ignored_messages=[
+        "Collapsing a non-contiguous coordinate."])
+    def test_weights_sum_to_zero(self):
+        """Test all x-y slices have zero weight in the same index.
+           This case corresponds to the case when all the input fields are
+           all masked in the same place."""
+        self.spatial_weights_cube.data[:, :, :, 0] = 0
+        expected_result = np.array([[[[0, 0, 0.2],
+                                      [0, 0, 0.2]],
+                                     [[0, 0, 0.2],
+                                      [0, 0, 0.2]]],
+                                    [[[0, 0, 0.5],
+                                      [0, 0, 0.5]],
+                                     [[0, 0, 0.5],
+                                      [0, 0, 0.5]]],
+                                    [[[0, 1.0, 0.3],
+                                      [0, 1.0, 0.3]],
+                                     [[0, 1.0, 0.3],
+                                      [0, 1.0, 0.3]]]],
+                                   dtype=np.float32)
+        result = self.plugin.normalised_masked_weights(
+            self.spatial_weights_cube, "forecast_reference_time")
+        self.assertArrayAlmostEqual(result.data, expected_result)
+        self.assertEqual(result.metadata, self.spatial_weights_cube.metadata)
+
+
+class Test_process(IrisTest):
+    """Test process method"""
```

Comment:
A lot of these tests are duplicates of the functions above.  Please could you consider reducing the duplication and only testing things which are specifically added by the "process" function (eg checking for appropriate output metadata / dimensions)?

Reply:
This duplication came about is by testing the process method for different shapes of array etc, and then noticing that other functions weren't quite working correctly, so I went back and added extra unit tests to those functions.

I was also trying to document clearly what the function does in different situations, by building up the different cases with different levels of fuzziness and different linear weights to check that the process function produces sensible weights in all the different combinations of how this plugin could be called. I didn't think it was a bad ideas to show how all the different functions worked together in the different cases.